finat/piola_mapped.py

```diff
@@ -0,0 +1,84 @@
+import numpy
+from gem import Literal
+
+
+def determinant(A):
+    """Return the determinant of A"""
+    n = A.shape[0]
+    if n == 0:
+        return 1
+    elif n == 1:
+        return A[0, 0]
+    elif n == 2:
+        return A[0, 0] * A[1, 1] - A[0, 1] * A[1, 0]
+    else:
+        detA = A[0, 0] * determinant(A[1:, 1:])
+        cols = numpy.ones(A.shape[1], dtype=bool)
+        for j in range(1, n):
+            cols[j] = False
+            detA += (-1)**j * A[0, j] * determinant(A[1:][:, cols])
+            cols[j] = True
+        return detA
+
+
+def adjugate(A):
+    """Return the adjugate matrix of A"""
+    A = numpy.asarray(A)
+    C = numpy.zeros_like(A)
+    rows = numpy.ones(A.shape[0], dtype=bool)
+    cols = numpy.ones(A.shape[1], dtype=bool)
+    for i in range(A.shape[0]):
+        rows[i] = False
+        for j in range(A.shape[1]):
+            cols[j] = False
+            C[j, i] = (-1)**(i+j)*determinant(A[rows, :][:, cols])
+            cols[j] = True
+        rows[i] = True
+    return C
+
+
+def piola_inverse(fiat_cell, J, detJ):
```

Comment:
The inverse of something is adjugate over determinant, so the inverse of the  Piola map is just the adjugate of the Jacobian since the determinants cancel?